xds/src/main/java/io/grpc/xds/FilterChainMatchingHandler.java

```diff
@@ -0,0 +1,338 @@
+import io.grpc.netty.InternalProtocolNegotiationEvent;
+import io.grpc.netty.InternalProtocolNegotiator.ProtocolNegotiator;
+import io.grpc.netty.ProtocolNegotiationEvent;
+import io.grpc.xds.EnvoyServerProtoData.CidrRange;
+import io.grpc.xds.EnvoyServerProtoData.ConnectionSourceType;
+import io.grpc.xds.EnvoyServerProtoData.FilterChain;
+import io.grpc.xds.EnvoyServerProtoData.FilterChainMatch;
+import io.grpc.xds.XdsServerWrapper.ServerRoutingConfig;
+import io.grpc.xds.internal.Matchers.CidrMatcher;
+import io.grpc.xds.internal.sds.SdsProtocolNegotiators.ServerSdsHandler;
+import io.grpc.xds.internal.sds.SslContextProviderSupplier;
+import io.netty.channel.ChannelHandler;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import java.net.Inet6Address;
+import java.net.InetAddress;
+import java.net.InetSocketAddress;
+import java.security.cert.CertStoreException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+
+/**
+ * Handles L4 filter chain match for the connection based on the xds configuration.
+ * */
+public final class FilterChainMatchingHandler extends ChannelInboundHandlerAdapter {
+  private final GrpcHttp2ConnectionHandler grpcHandler;
+  private final FilterChainSelector selector;
+  @Nullable
+  private final ProtocolNegotiator fallbackProtocolNegotiator;
+
+  /**
+   * Selects the filter chain using the selector configuration.
+   * */
+  public FilterChainMatchingHandler(
+          GrpcHttp2ConnectionHandler grpcHandler, FilterChainSelector selector,
+          @Nullable ProtocolNegotiator fallbackProtocolNegotiator) {
+    this.grpcHandler = checkNotNull(grpcHandler, "grpcHandler");
+    this.selector = checkNotNull(selector, "selector");
+    this.fallbackProtocolNegotiator = fallbackProtocolNegotiator;
+  }
+
+  @Override
+  public void userEventTriggered(ChannelHandlerContext ctx, Object evt) throws Exception {
+    if (!(evt instanceof ProtocolNegotiationEvent)) {
+      super.userEventTriggered(ctx, evt);
+      return;
+    }
+    FilteredConfig config = selector.select(
+            (InetSocketAddress) ctx.channel().localAddress(),
+            (InetSocketAddress) ctx.channel().remoteAddress());
+    if (config == null) {
+      throw new IllegalStateException("not filter chain");
+    }
+    ProtocolNegotiationEvent pne = InternalProtocolNegotiationEvent.getDefault();
+    Attributes attr = InternalProtocolNegotiationEvent.getAttributes(pne)
+            .toBuilder().set(ATTR_SERVER_ROUTING_CONFIG, config.routingConfig).build();
+    pne = InternalProtocolNegotiationEvent.withAttributes(pne, attr);
+    SslContextProviderSupplier sslContextProviderSupplier = config.sslContextProviderSupplier;
+
+    ChannelHandler handler;
+    if (sslContextProviderSupplier == null) {
+      if (fallbackProtocolNegotiator == null) {
```

Comment:
Ideally `fallbackProtocolNegotiator` should never be null if you follow how it is supplied via `XdsServerCredentials`. So we can add the `checkNotNull` checks everywhere it is passed down and skip these checks.

Reply:
Just leave it as is.  This handler doesn't need to be changed (mostly). If we need a `checkNotNull` let's fix it after.